### ISSUE TITLE
payara/server-full:5.2020.4-jdk11 retries servlet actions longer than 30 sec

### ISSUE THREAD

**Issue by dbbernstein** [payara/Payara#4954]
# Description #
----------

Intentionally long-running servlet request from browser is repeatedly retried once it exceeds 30 seconds, resulting in multiple overlapping operations.

## Expected Outcome

Servlet request should run to completion without retry even if it exceeds 30 seconds. Payara 5.201 works this way.

In any case, if there is a retry, there should by default be a log message in `server.log` explaining what happened and how to change it, as well as how to adjust any relevant logging. Presumably the timeout and number of retries are configurable. If this is intentional behavior, then the severity of the log message should be configurable with default `Level.Warning`.

## Current Outcome

Using Dockerized Payara image `payara/server-full:5.2020.4-jdk11`, once the servlet operation exceeds 30 seconds, it is retried in parallel with any previous attempts. Payara 5.201 does not work this way and I was unable to find documentation of an intended change so this is apparently a regression.

## Steps to reproduce (Only for bug reports)

1. Compile the attached simple single-file servlet `BarServlet`.
1. Package the servlet into a WAR including `web.xml` that attaches the servlet to a desired URL. Usable snippet below.
1. Deploy the WAR to a Dockerized Payara image `payara/server-full:5.2020.4-jdk11`.
1. Request the URL from a browser (I've been using Safari, but have also seen Chrome fail). Behavior observed when browser and Dockerized Payara running on same computer.

### Samples

See attached extremely simple servlet Java code `BarServlet.java` to be placed in package `com.foo`. (Suffixed .TXT to allow Github upload.) See also `web.xml` snippet:

```
	<servlet>	
		<servlet-name>bar</servlet-name>
		<servlet-class>com.foo.BarServlet</servlet-class>
	</servlet>
	<servlet-mapping>
		<servlet-name>bar</servlet-name>
		<url-pattern>/bar</url-pattern>
	</servlet-mapping>
```	

## Context (Optional)

IRL, the goal is to run a suite of tests that are idempotent, but **not thread-safe**, because they each expect to perform database operations without interference on the same data. So, overlapping execution of multiple instances of the test suite causes spurious test failure. Rearchitecture of the test suite would be required to work around this change in behavior from Payara 5.201, so this is blocking the upgrade to 5.2020.4-jdk11, and the motivating factor for the upgrade was HTTP/2 capability without extra work. In this test case, the simple servlet gets run repeatedly while the actual tests are only run three times, with the final one being returned; I don't know why this difference exists, but they both share the change in behavior from 5.201 to 5.2020.4-jdk11.

Searching, I was unable to find any documentation for this change in behavior from Payara 5.201. If it exists, an RTFM pointer would be appreciated and would presumably unblock me. In any case, I think that by default some message should be logged. At one point I thought that perhaps `org.eclipse.microprofile.faulttolerance` `Timeout` and/or `Retry` were the cause, but I could find no evidence of them having been attached to relevant classes or methods or somehow otherwise applied.

## Environment ##

- **Payara Version**: 5.2020.4-jdk11
- **Edition**: Full 
- **JDK Version**: Payara running with Zulu JDK 11, code compiled Zulu JDK 8 
- **Operating System**: Mac 10.15.6, but Payara is Dockerized, see above 
- **Database**: problem manifest without database 
- **Browser**: Safari 14.0 requesting HTTP/2, but also saw behavior using Chrome 86.0.4240.75

[BarServlet.java.txt](https://github.com/payara/Payara/files/5387115/BarServlet.java.txt)

**Comment by rdebusscher** [comment]
Hi David,

I was unable to reproduce this. This is what I have done

- Created a web application with the BarServlet you attached.
- Started Payara Docker Image `docker run -p 4848:4848 -p 8080:8080 -d payara/server-full:5.2020.4-jdk11`
- Deployed the test application
- Verified with Safari, Chrome and Firefox
- In the browser I get a response after 45 secs and I see in the log only once the sequence  ENTER - Waiting - DONE - EXIT.

- You mention HTTP/2, so are you using a HTTPS connection?
- Are we sure the browser is not resending the request (as according to the HTTP spec, the client should resend the request when timeout in certain conditions) or due to a plugin? Do you have the same issue with curl?

Best Regards
Rudy

**Comment by dbbernstein** [comment]
Hi Rudy,

Well, I believe you and thanks for looking at it. When I run the test case I gave you in a super-isolated context (nothing else deployed), it does in fact behave as I expected and you observed. I did spend quite a while re-investigating this after you responded, but sadly discovered no useful additional information.
 
When I run the servlet in the context of my real production application (a few WAR files, where it's embedded in the main/largest one), then it works that way with 5.201 but for 5.2020.4-jdk11 it breaks with the previously observed repeated invocation. I've tested this multiple times changing only whether the payara Docker image is payara/server-full:5.2020.4-jdk11 or payara/server-full:5.201 and that single change was enough to flip the behavior.

Unfortunately, at this point, I am completely stumped about how to narrow my real app down to a test case that would trigger the same behavior. I have just rewritten my code to return quickly, so this is no longer an immediate issue for me, but I do believe that something has changed in Payara to attempt a retry every 30 seconds for certain long-running servlet request. I realize that I can't produce a usable test case so the issue should be closed, which I'll do now. Sorry.

In case it helps any possible future investigation:
* Yes, I was using HTTPS/TLS. I was using my real certificate and put an exception in my browser to allow it on my development machine.
* I believe I wasn't using HTTP/2, even though migration to out-of-the-box HTTP/2 capability both client and server are a major motivation for this update for me. I'm pretty sure this is so because HttpServletRequest.newPushBuilder() returned null and there was no evidence of an HTTP/2 request in the headers.
* I don't think that the browser was retrying but I suppose could be missing it somehow.

Thanks again,
David
